File: Visual Studio LIghtswitch 2012/Chapter13/Chapter13/ApressExtensionCS/ApressExtensionCS/ApressExtensionCS.Server/DataSources/WindowsEventLog.cs
//You can use and redistribute the code from this book (and sample application) for non-commercial and 
//most commercial purposes as long as you acknowledge the source and authorship. 
//You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
//The acknowledgment should include author, title, publisher, and ISBN. 
//An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2012 by Tim Leung, published by Apress, ISBN 9781430250715".

using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.ServiceModel.DomainServices.Server;
using System.Diagnostics;

namespace ApressExtensionCS.DataSources
{
    public class WindowsEventLog : DomainService
    {

        private string _serverName;
        public override void Initialize(DomainServiceContext context)
        {
            base.Initialize(context);
        }

        public override bool Submit(ChangeSet changeSet)
        {
            Boolean baseResult = base.Submit(changeSet);
            return true;
        }

        protected override int Count<T>(IQueryable<T> query)
        {
            return query.Count();
        }

        [Query(IsDefault = true)]                                              
        public IQueryable<LogEntry> GetEventEntries()
        {

            int idCount = 0;
            List<LogEntry> eventLogs = new List<LogEntry>();
            LogSource logSource = default(LogSource);

            foreach (var eventSource in EventLog.GetEventLogs("."))
            {
                logSource = new LogSource();
                logSource.SourceName = eventSource.Log;

                try
                {
                    foreach (System.Diagnostics.EventLogEntry eventEntry 
                       in eventSource.Entries)
                    {
                        LogEntry newEntry = new LogEntry();
                        newEntry.LogEntryID = idCount;

                        newEntry.EventDateTime = eventEntry.TimeWritten;
                        newEntry.Message = eventEntry.Message;
                        newEntry.Message = eventEntry.Source;
                        newEntry.SourceName = eventSource.Log;
                        newEntry.EventSource = logSource;
                        eventLogs.Add(newEntry);

                        idCount += 1;
                        if (idCount > 200)
                        {
                            break;  
                        }
                    }
                }
                catch (System.Security.SecurityException ex)
                {
                    //User doesn't have access to view the log
                    //Move onto the next log
                }
            }

            return eventLogs.AsQueryable();
        }

        [Query(IsDefault = true)]                                          
        public IQueryable<LogSource> GetEventLogTypes()
        {

            List<LogSource> eventLogs = new List<LogSource>();

            foreach (var elEventEntry in
               System.Diagnostics.EventLog.GetEventLogs())
            {
                LogSource event1 = new LogSource();
                event1.SourceName = elEventEntry.Log;
                eventLogs.Add(event1);
            }

            return eventLogs.AsQueryable();
        }

        public void InsertLogEntry(LogEntry entry)
        {
            try
            {
                using (System.Diagnostics.EventLog applicationLog =  
                   new System.Diagnostics.EventLog("Application", "."))
                {
                    applicationLog.Source = "Application";
                    applicationLog.WriteEntry(
                      entry.Message, EventLogEntryType.Warning);
                }

            }
            catch (Exception ex)
            {
                throw new Exception(
                  "Error writing Event Log Entry" + ex.Message);
            }
        }

        public void UpdateLogEntry(LogEntry entry)
        {}

        public void DeleteLogEntry(LogEntry entry)
        {}
    }
}

File: Visual Studio LIghtswitch 2012/Chapter13/Chapter13/ApressExtensionVB/ApressExtensionVB/ApressExtensionVB.Server/DataSources/WindowsEventLog.vb
'You can use and redistribute the code from this book (and sample application) for non-commercial and 
'most commercial purposes as long as you acknowledge the source and authorship. 
'You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
'The acknowledgment should include author, title, publisher, and ISBN. 
'An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2012 by Tim Leung, published by Apress, ISBN 9781430250715".

'Listing 13-25. Domain Service Code for Accessing Data
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.ComponentModel.DataAnnotations
Imports System.Linq
Imports System.ServiceModel.DomainServices.Server

Imports System.Configuration
Imports System.Web.Configuration
Imports System.Diagnostics.EventLog

Namespace DataSources

    <Description("Enter your server path.")> _
    Public Class WindowsEventLog
        Inherits DomainService

        Private _serverName As String

        Public Overrides Sub Initialize(context As DomainServiceContext)
            MyBase.Initialize(context)
        End Sub

        Public Overrides Function Submit(changeSet As ChangeSet) As Boolean
            Dim baseResult As [Boolean] = MyBase.Submit(changeSet)
            Return True
        End Function

#Region "Queries"

        Protected Overrides Function Count(Of T)(query As IQueryable(Of T)) As Integer
            Return query.Count()
        End Function

        <Query(IsDefault:=True)> _
        Public Function GetEventEntries() As IQueryable(Of LogEntry)

            Dim idCount As Integer = 0
            Dim eventLogs As New List(Of LogEntry)()
            Dim logSource As LogSource

            For Each eventSource In EventLog.GetEventLogs(".")

                logSource = New LogSource
                logSource.SourceName = eventSource.Log

                Try
                    For Each eventEntry As System.Diagnostics.EventLogEntry
                       In eventSource.Entries
                        Dim newEntry As New LogEntry
                        newEntry.LogEntryID = idCount

                        newEntry.EventDateTime = eventEntry.TimeWritten
                        newEntry.Message = eventEntry.Message
                        newEntry.Message = eventEntry.Source
                        newEntry.SourceName = eventSource.Log
                        newEntry.EventSource = logSource
                        eventLogs.Add(newEntry)

                        idCount += 1
                        If idCount > 200 Then
                            Exit For
                        End If

                    Next

                Catch ex As System.Security.SecurityException
                    'User doesn't have access to view the log
                    'Move onto the next log
                End Try
            Next

            Return eventLogs.AsQueryable
        End Function

        <Query(IsDefault:=True)> _
        Public Function GetEventLogTypes() As IQueryable(Of LogSource)

            Dim eventLogs As New List(Of LogSource)()
            For Each elEventEntry In System.Diagnostics.EventLog.GetEventLogs

                Dim event1 As New LogSource
                event1.SourceName = elEventEntry.Log
                eventLogs.Add(event1)
            Next

            Return eventLogs.AsQueryable
        End Function

        Public Sub InsertLogEntry(entry As LogEntry)
            Try
                Using applicationLog As New System.Diagnostics.EventLog("Application", ".")
                    applicationLog.Source = "Application"
                    applicationLog.WriteEntry(
                       entry.Message, EventLogEntryType.Warning)
                End Using

            Catch ex As Exception
                Throw New Exception("Error writing Event Log Entry" & ex.Message)
            End Try
        End Sub

        Public Sub UpdateLogEntry(entry As LogEntry)
        End Sub

        Public Sub DeleteLogEntry(entry As LogEntry)
        End Sub

#End Region

    End Class

End Namespace
